src/RestApi/StoreApi/Utilities/ReserveStock.php

```diff
@@ -0,0 +1,141 @@
+<?php
+/**
+ * Helper class to handle product stock reservation.
+ *
+ * @package WooCommerce/Blocks
+ */
+
+namespace Automattic\WooCommerce\Blocks\RestApi\StoreApi\Utilities;
+
+defined( 'ABSPATH' ) || exit;
+
+use \WP_Error;
+
+/**
+ * Stock Reservation class.
+ */
+class ReserveStock {
+	/**
+	 * Put a temporary hold on stock for an order if enough is available.
+	 *
+	 * @throws RestException Exception when stock cannot be reserved.
+	 * @param \WC_Order $order Order object.
+	 * @return bool|WP_Error
+	 */
+	public function reserve_stock_for_order( \WC_Order $order ) {
+		$stock_to_reserve = [];
+		$items            = array_filter(
+			$order->get_items(),
+			function( $item ) {
+				return $item->is_type( 'line_item' ) && $item->get_product() instanceof \WC_Product;
+			}
+		);
+
+		foreach ( $items as $item ) {
+			$product = $item->get_product();
+
+			if ( ! $product->is_in_stock() ) {
+				return new WP_Error(
+					'product_out_of_stock',
+					sprintf(
+						/* translators: %s: product name */
+						__( '%s is out of stock and cannot be purchased.', 'woo-gutenberg-products-block' ),
+						$product->get_name()
+					),
+					[ 'status' => 403 ]
+				);
+			}
```

Comment:
I see the value here in using WP_Error because it allows for easier return of custom data. However, I'd love it if we could move to using exceptions more. We could create our own named exceptions.  For instance you could have a `OutOfStockException` (and for use later a `NotEnoughStockException`) that receives the product name and status code to return.  I think error handling is much more straightforward with exceptions.

With that said, I don't want to block this pull. If you agree, just comment and I'll create an issue for considering implementing custom exceptions at some point.

Reply:
Maybe worth a shot. Only concern would be over-engineering when something can be handled by a single existing error class. Might just be set in my ways.